### PR TITLE
Artery transport shutdown improvements

### PR DESCRIPTION
**Pull request by johanandren**
Refs #21352

### CODE REVIEW

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala [hunk 1]
```diff
@@ -698,6 +701,24 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
     }
   }
 
+  // set the future that completes when the current stream for a given name completes
+  @tailrec
+  private def updateStreamCompletion(streamName: String, streamCompleted: Future[Done]): Unit = {
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: what is the `streamName` used for? can't we use a `Set`?

Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: ah, it's for replacing, ofc

Comment (johanandren) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: Yeah, maybe we could just add _all_ of them in a set, the restarted ones will have completed as well in the end, no?

File: akka-remote/src/main/scala/akka/remote/artery/Association.scala [hunk 2]
```diff
@@ -427,6 +428,7 @@ private[remote] class Association(
 
   private def attachStreamRestart(streamName: String, streamCompleted: Future[Done], restart: Throwable ⇒ Unit): Unit = {
     implicit val ec = materializer.executionContext
+    updateStreamCompletion(streamName, streamCompleted)
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/Association.scala]: In https://github.com/akka/akka/pull/21356 I have added the queueIndex as parameter here so then we could use that as key. The purpose of the `streamName` was only for logging.

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala (hunk 1 above)
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: if we don't want to use the `streamName` as key we could pass in the stream id to the restart method
```
  private val controlStreamId = 1
  private val ordinaryStreamId = 2
  private val largeStreamId = 3
```

**Comment by patriknw** [comment]
looking good, with some thoughts about the `streamName`

File: akka-remote/src/main/scala/akka/remote/artery/Association.scala [hunk 3]
```diff
@@ -446,6 +448,21 @@ private[remote] class Association(
...
+  private def updateStreamCompletion(streamName: String, streamCompleted: Future[Done]): Unit = {
+    val prev = streamCompletions.get()
+    if (!streamCompletions.compareAndSet(prev, prev + (streamName → streamCompleted))) {
+      updateStreamCompletion(streamName, streamCompleted)
+    }
+  }
+
+  // for orderly shutdown purposes, can not be trusted except for during shutdown as streams may restart
+  def streamsCompleted: Future[Done] = {
+    implicit val ec = materializer.executionContext
```
Comment (johanandren) [inline akka-remote/src/main/scala/akka/remote/artery/Association.scala]: is it dangerous to run this on the materializer EC perhaps?

Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/Association.scala]: I think it's fine. everything has to complete anyway

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala (hunk 1 above)
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: but then the set will grow over time if there are many restarts

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala [hunk 4]
```diff
@@ -670,9 +670,12 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
         flushingPromise.future
       }
     implicit val ec = remoteDispatcher
-    flushing.recover { case _ ⇒ Done }.map { _ ⇒
-      killSwitch.shutdown()
 
+    for {
+      _ ← flushing.recover { case _ ⇒ Done }
+      _ = killSwitch.shutdown()
+      _ ← streamsCompleted
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: what happens if the stream futures are completed with failure for some reason?
either we just use an `onComplete` callback instead, or we have to `recover` each one?

Comment (johanandren) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: True. Do we really want to silence any errors though? Collect and log or something like that?

Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: you can change the `attachStreamRestart` methods to do logging of errors also during shutdown, they are ignored there in `case _ if isShutdown`. Then we have all such logging in one place.

Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: @johanandren the `recover` is also missing

Comment (johanandren) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: I put the recover on the futures where failure would actually happen, this is the aggregate, I don't think Future.sequence will ever fail unless the sequenced failures does.

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala [hunk 5]
```diff
@@ -663,8 +664,11 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
 
   private def attachStreamRestart(streamName: String, streamCompleted: Future[Done], restart: () ⇒ Unit): Unit = {
     implicit val ec = materializer.executionContext
+    updateStreamCompletion(streamName, streamCompleted.recover { case _ ⇒ Done })
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: ah, this is the recover I was looking for, thanks

File: akka-remote/src/main/scala/akka/remote/artery/Association.scala [hunk 6]
```diff
@@ -534,8 +535,11 @@ private[remote] class Association(
     }
 
     implicit val ec = materializer.executionContext
+    updateStreamCompletion(streamName, streamCompleted.recover { case _ ⇒ Done })
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/Association.scala]: and here, all good then

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala (hunk 4 above)
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: thanks, missed that

**Pull request merged**